### ISSUE TITLE
eval loss/acc指标不太对

### ISSUE THREAD

**Issue by eshoyuan** [modelscope/ms-swift#2489]
<img width="786" alt="image" src="https://github.com/user-attachments/assets/2f166e60-f958-4688-b572-3937492cdde8">

![image](https://github.com/user-attachments/assets/cfdd7247-822d-4f7b-b26f-0d267dba35a5)

![image](https://github.com/user-attachments/assets/16b511f4-ad8a-413c-ac6c-6f8c729dc01f)

![image](https://github.com/user-attachments/assets/69ca0344-3a0c-4415-8813-b51411da13a3)

我只训练了两个epoch, 至少第一个epoch上, eval loss/acc的趋势应该和train上一样的, 因为都是没有见过的数据. eval的数据集是用的默认0.01的划分, 不存在domain不一样的情况. 请问一下有什么可能的原因吗? 

```export NCCL_IB_DISABLE=0
NPROC_PER_NODE=8 \
MASTER_PORT=29500 \
USE_HF=1 \
swift sft \
    --model_id_or_path deepseek-ai/deepseek-coder-6.7b-base \
    --sft_type full \
    --template_type AUTO \
    --dtype AUTO \
    --output_dir output \
    --dataset $DATASET \
    --num_train_epochs 2 \
    --max_length 2048 \
    --check_dataset_strategy warning \
    --gradient_checkpointing true \
    --batch_size 12 \
    --weight_decay 0.1 \
    --learning_rate 5e-5 \
    --gradient_accumulation_steps 4 \
    --warmup_steps 15 \
    --eval_steps 50 \
    --save_only_model true \
    --save_total_limit 2 \
    --logging_steps 10 \
    --use_flash_attn true \
    --preprocess_num_proc 8 \
    --deepspeed 'default-zero2'```

**Comment by eshoyuan** [comment]
版本 transformers                      4.46.3
ms-swift                          2.6.0.post1

**Comment by eshoyuan** [comment]
补充一下, 刚刚发现了了一个可能的原因.
```
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
NPROC_PER_NODE=8 \
MASTER_PORT=29500 \
USE_HF=1 \
swift sft \
    --model_id_or_path deepseek-ai/deepseek-coder-6.7b-base \
    --sft_type full \
    --template_type AUTO \
    --dtype AUTO \
    --output_dir output \
    --dataset 38k.jsonl \
    --num_train_epochs 2 \
    --max_length 2048 \
    --check_dataset_strategy warning \
    --gradient_checkpointing true \
    --batch_size 12 \
    --weight_decay 0.1 \
    --learning_rate 5e-5 \
    --gradient_accumulation_steps 4 \
    --warmup_steps 15 \
    --eval_steps 50 \
    --save_only_model true \
    --save_total_limit 2 \
    --logging_steps 10 \
    --use_flash_attn true \
    --preprocess_num_proc 8 \
    --deepspeed 'default-zero2'
```

这是我的脚本, 按理来说, steps应该是38k*2/8/12/4的, 大概不到200steps, 但是运行的max steps缺为1578. 差了8倍. 曲线结果也比较符合, 在0-100之间, eval和train的loss是一致的.

**Comment by eshoyuan** [comment]
似乎是preprocess_num_proc导致的, 很奇怪

**Comment by Jintao-Huang** [comment]
我看看

**Comment by eshoyuan** [comment]
> 我看看

辛苦了, https://github.com/eshoyuan/Swift-train/blob/main/scripts/train.sh这个repo下是我的脚本以及用到的数据.

**Comment by Jintao-Huang** [comment]
main分支已修复
